docs/architecture/2024-10-23_osquery_refactor.md

```diff
@@ -0,0 +1,92 @@
+
+Therefore, we will want to refactor our current implementation with the following goals in mind:
+
+1. We want one osquery extension per osquery instance
+2. We want to be able to add new osquery instances without requiring a full launcher restart
+3. We do not want onerously complex setup for the osquery instances or their extensions, since our osquery process management and osquery data exchange is already a complicated system
+
+## Decision
+
+To simplify the creation of multiple osquery instances, and to make it possible to increase the number of osquery instances running without performing a full launcher restart, we will move the osquery extension ownership to the osquery instance.
+
+This will also give us the opportunity to make the separation of responsibility between the runner and the instance clearer. The instance will create its own extension, extension server, etc. The runner will be responsible for the instances' state. This means the osquery instance stats should be owned by the runner, since the runner already has the responsibility of updating them.
+
+```mermaid
+---
+title: Refactored (one osquery instance)
+---
+erDiagram
+    RUNGROUP ||--|| OSQUERYRUNNER : runs
+    OSQUERYRUNNER ||--|| OSQUERYINSTANCE: runs
+    OSQUERYINSTANCE ||--|| EXTENSION : owns
+    EXTENSION ||--|| JSONRPCCLIENT : uses
+    OSQUERYINSTANCE ||--|{ EXTENSIONSERVER : runs
+    OSQUERYINSTANCE ||--|| EXTENSIONCLIENT : runs
+    EXTENSIONSERVER ||--|| EXTENSION : forwards
+    OSQUERYRUNNER ||--|| STATS: updates
+```
+
+The below diagram shows the desired future state -- it is identical to the above diagram except for the relationship cardinalities (e.g. `OSQUERYRUNNER` runs 0 or more of `OSQUERYINSTANCE`, rather than exactly 1 `OSQUERYINSTANCE`).
+
+```mermaid
+---
+title: Future state (0-n osquery instances)
+---
+erDiagram
+    RUNGROUP ||--|| OSQUERYRUNNER : runs
+    OSQUERYRUNNER ||--o{ OSQUERYINSTANCE: runs
+    OSQUERYINSTANCE ||--|| EXTENSION : owns
+    EXTENSION }o--|| JSONRPCCLIENT : uses
+    OSQUERYINSTANCE ||--|{ EXTENSIONSERVER : runs
+    OSQUERYINSTANCE ||--|| EXTENSIONCLIENT : runs
+    EXTENSIONSERVER ||--|| EXTENSION : forwards
+    OSQUERYRUNNER ||--o{ STATS: updates
+```
+
+## Consequences
+
+The osquery instance will now be responsible for managing extension health -- if the extension exits, the instance will need to take corrective action. (This is potentially a better outcome than before, since now extension shutdown can be remediated without a full launcher restart.)
```

Comment:
> The osquery instance will now be responsible for managing extension health

Do you mean the go actor sometimes called an osquery instance, or do you mean the osquery process will launch extension under it?